src/evo/netinfo.h

```diff
@@ -42,6 +44,91 @@ constexpr std::string_view NISToString(const NetInfoStatus code)
     assert(false);
 }
 
+class NetInfoEntry
+{
+public:
+    enum NetInfoType : uint8_t {
+        Service = 0x01,
+        Invalid = 0xff
+    };
+
+private:
+    uint8_t m_type{NetInfoType::Invalid};
+    std::variant<std::monostate, CService> m_data{std::monostate{}};
+
+public:
+    NetInfoEntry() = default;
+    NetInfoEntry(const CService& service)
+    {
+        if (!service.IsValid()) return;
+        m_type = NetInfoType::Service;
+        m_data = service;
+    }
+
+    ~NetInfoEntry() = default;
+
+    bool operator<(const NetInfoEntry& rhs) const;
+    bool operator==(const NetInfoEntry& rhs) const;
+    bool operator!=(const NetInfoEntry& rhs) const { return !(*this == rhs); }
+
+    template <typename Stream>
+    void Serialize(Stream& s) const
+    {
+        if (const auto* data_ptr{std::get_if<CService>(&m_data)};
+            m_type == NetInfoType::Service && data_ptr && data_ptr->IsValid()) {
+            s << m_type << *data_ptr;
+        } else {
+            s << NetInfoType::Invalid;
+        }
+    }
+
+    void Serialize(CSizeComputer& s) const
+    {
+        auto size = ::GetSerializeSize(uint8_t{}, s.GetVersion());
+        if (m_type == NetInfoType::Service) {
+            size += ::GetSerializeSize(CService{}, s.GetVersion());
+        }
+        s.seek(size);
+    }
+
+    template <typename Stream>
+    void Unserialize(Stream& s)
+    {
+        s >> m_type;
+        if (m_type == NetInfoType::Service) {
+            m_data = CService{};
+            try {
+                CService& service{std::get<CService>(m_data)};
+                s >> service;
+                if (!service.IsValid()) {
+                    throw std::ios_base::failure("Invalid CService");
```

Comment:
oh no, throwing exception is very expensive operation. Can you refactor this code to avoid throwing exception?

Reply:
Replaced exception with `Clear()`. Resolved in latest push!